core/src/main/scala/kafka/server/KafkaApis.scala

```diff
@@ -1417,21 +1426,35 @@ class KafkaApis(val requestChannel: RequestChannel,
     val transactionalId = addPartitionsToTxnRequest.transactionalId
     val partitionsToAdd = addPartitionsToTxnRequest.partitions
 
-    // Send response callback
-    def sendResponseCallback(error: Errors): Unit = {
-      def createResponse(throttleTimeMs: Int): AbstractResponse = {
-        val responseBody: AddPartitionsToTxnResponse = new AddPartitionsToTxnResponse(throttleTimeMs, error)
-        trace(s"Completed $transactionalId's AddPartitionsToTxnRequest with partitions $partitionsToAdd: errors: $error from client ${request.header.clientId}")
-        responseBody
+    if(!authorize(request.session, Write, new Resource(ProducerTransactionalId, transactionalId)))
+      sendResponseMaybeThrottle(request, (throttleTimeMs: Int) => new InitPidResponse(throttleTimeMs, Errors.TRANSACTIONAL_ID_AUTHORIZATION_FAILED))
+    else {
+      val (_, unAuthorized) = partitionsToAdd.asScala.partition {
+        tp => authorize(request.session, Write, new Resource(Topic, tp.topic))
```

Comment:
One of the tricky aspects of topic authorization is that we try to avoid letting the APIs leak the topic names to users without Describe access. So usually what we do is first filter out all topics which don't have Describe access and return an UNKNOWN_TOPIC_OR_PARTITION error for them. Then we check Write permission and so on.